behaviors/services/trie.md

```diff
@@ -0,0 +1,97 @@
+* A Merkle Proof is a Hash Chain comprising of a single path within the Merkle Trie (Hash Tree). As such, it is merely a representation of a string of interconnected nodes contained in the Merkle Trie. 
+    * The proof corresponds to a path in the trie.
+    * The proof corresponds to a state key prefix, which may be a full state key or a partial state key.
+    * In a Merkle Proof array, every pair of adjacent nodes are parent and child the Merkle Tree 
+    * A Merkle Proof is valid when: 
+        * The first node’s hashcode is identical to the requested block’s Merkle Root 
+        * Each child node’s hashcode is identical to the parent reference to it.
+        * The key prefix assembled from all nodes up until the one before last in the proof is a prefix of the requested path
+    * A merkle proof which is valid may indicate a missing key or indicate the validated hashcode of the value associated with the key  
+        * Missing Key indication - if the proof ends with a branch node and the digit in the requested path does not reference any child node, or, if the final node indicates a divergence from the requested key in a way that negates any possibility that the requested key may be present in the trie.
+        * Present Key with value hash indication - if the proof ends with a lead node and the assembled key prefix is identical to the requested path
+* Prefixes are represented as hexadecimal digits in base64 encoding.
+* Zero values are never represented in the trie because they are not part of the state snapthot
+* Each node in a Merkle Trie is addressable via it’s hash-code (cryptographic digest) in a __content-addressed__ scheme - addresses of nodes are derived from their content as a cryptographic hashcode digest.  
+* Each block height represents a (possibly) distinct state in it’s virtual chain. For any block height  with a full Merkle trie and a Merkle Trie root corresponding to the matching state image.
+
+## Hash Function 
+TBD
+## Serialization 
+TBD - decide between
+ * Option 1 - custom - specify optimized serialization… ?
+ * Option 2 - Membuffers - possible issues with size
+ * Option 3 - RLP - for code simplicity and gas savings on EVM
+ 
+## Trie Key limitations
+### Uniform Key Length 
+All trie keys are of a uniform length. This means branch nodes and value nodes are mutually exclusive as there may not be two keys where one is contained in the other.
+   * Simplify trie structure by forcing values to be only in leaf nodes
+   * This means clients interpreting proofs must be aware of the uniform key length and enforce this knowledge when evaluating proofs
+   * This requirement depends on our state addressing scheme. See state storage requirements
+## Node types and structure 
+### Common properties
+All nodes have a prefix part. A Prefix part is the current node’s contribution to the key prefix represented by a path in the trie 
```

Comment:
prefix maybe empty

Reply:
In a node a prefix may be empty. this is not written. do you suggest we add this? this section is about keys in their entirety, not partial prefixes so I'm not sure this is the right place for it. is this what you meant?